### ISSUE TITLE
[BUG]: invalid plpsql function parsing

### ISSUE THREAD

**Issue by quenbyako** [pganalyze/libpg_query#85]
found some strange bug: declaring results to out arguments throws strange error:

when i calling `ParsePlPgSql` method with this sql query:
``` sql 
CREATE FUNCTION public.somefunc(OUT _result uuid[])
    RETURNS uuid[]
    LANGUAGE 'plpgsql'

AS $BODY$
DECLARE
	active_on_to_date uuid[];
BEGIN
_result := ARRAY( SELECT some_id FROM some_table);
END;
$BODY$;
```
i got really funny issue
```
"_result" is not a known variable
```

i used official go wrapper v1.0.2, but it looks like this issue happens on exactly this base library

**Comment by lfittl** [comment]
@quenbyako Thanks for the report - is this a new problem with the v1.0.2 release of the Go library, or already present in v1.0.1?

**Comment by quenbyako** [comment]
@lfittl wow, that's fast! thanks for response :smiley: 

Well, i checked out to v1.0.1 commit and run my code again, so i got absolutely same error with additional warning:
```
pg_query_parse_plpgsql.c: In function ‘pg_query_parse_plpgsql’:
pg_query_parse_plpgsql.c:437:10: warning: ignoring return value of ‘asprintf’, declared with attribute warn_unused_result [-Wunused-result]
  437 |    (void)asprintf(&new_out, "%s%s,\n", result.plpgsql_funcs, func_json);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

this appears only in 1.0.1, fresh version doesn't print it. Maybe this warn is related just for go code, but i think it could be helpful to you)

**Comment by lfittl** [comment]
@quenbyako Thanks for checking - I released an update to the go library earlier in the week that touched the PL/pgSQL code, and I was worried that I introduced a breaking change (looks like I did not, as it was already broken before).

The warning is harmless and unrelated (and fixed in v1.0.2).

The problem with the function is I think a known issue of how we call the PL/pgSQL parser today (it hard codes a few variables), and will require some work to investigate and fix.

Whats the use case you are looking at pg_query_go for - is it important to have PL/pgSQL function parsing? (vs just query parsing)

**Comment by quenbyako** [comment]
Btw, i think that variables definitions checking is not a part of AST builder, cuz this is actually logic error instead syntax. So maybe, it's better to ignore this issue and leave this check to other parts of client code)

**Comment by quenbyako** [comment]
> Whats the use case you are looking at pg_query_go for - is it important to have PL/pgSQL function parsing? (vs just query parsing)

Yeah, basic parsing works perfectly well, and i already made a few tools with your lib (really good one, thanks), now i have tricky task to check does the function logic redeclared differently than previous version. diffs not work, cause let's be honest, in sql newline, additional space or tab doesn't make sense to code, so it's better to parse it and check ast than diff raw text)

**Comment by quenbyako** [comment]
@lfittl so, does this issue could be fixed in short time? if yes, it'll be really amazing

**Comment by lfittl** [comment]
I think the problem lies around this area of the code:

https://github.com/lfittl/libpg_query/blob/10-latest/src/pg_query_parse_plpgsql.c#L165

The problem is that whilst the regular query parser is fairly independent, the PL/pgSQL parser is more tied to the actual catalog tables being accessible (to validate data types, etc). So this is a complex problem to resolve, which may require mocking some of the code that attempts to verify types/etc.

I won't have time to address this in the near future (treat it as a known issue), but if you have time you could investigate this in the C code and provide a PR - I'd be happy to review.

**Comment by quenbyako** [comment]
sad, but i don't understand code on clang :cry: 

Okay, i got it, i'll try to ask our team to deal with this problem, maybe we'll try to contribute and fix this one)

Thanks for helping out!

**Comment by lfittl** [comment]
Fixed in https://github.com/pganalyze/libpg_query/commit/e9e4ba9ffcbac02bce6656ecabd22c2b54eca1b2
